### PR TITLE
49 improve the program entry point and main setup

### PR DESCRIPTION
**Pull request by Wuqiqi123**
1. Becase we are not using the QApplication, use Qt's signal/slot in the same thread is fine.
2. [cxxopts](https://github.com/jarro2783/cxxopts)(header only) is introduced for parsing command line parameter.

### CODE REVIEW

File: src/gobot/scene/scene_tree.cpp [hunk 1]
```diff
@@ -8,34 +8,90 @@
...
+
+void SceneTree::Finalize()
+{
+    MainLoop::Finalize();
+
+    if (root) {
+        root->SetTree(nullptr);
+        root->PropagateAfterExitTree();
+        Node::Delete(root);
+    }
```
Comment (StephenYu-R) [inline src/gobot/scene/scene_tree.cpp:38]: root = nullptr;

File: tests/scene/test_node.cpp [hunk 2]
```diff
@@ -8,15 +8,18 @@
...
 class TestNode : public testing::Test {
 protected:
     void SetUp() override {
         tree = gobot::SceneTree::New<gobot::SceneTree>();
+        tree->Initialize();
         node = gobot::Node::New<gobot::Node>();
     }
 
     void TearDown() override {
+        tree->Initialize();
```
Comment (StephenYu-R) [inline tests/scene/test_node.cpp:22]: tree->Finalize();

File: tests/scene/test_node_complex.cpp [hunk 3]
```diff
@@ -8,17 +8,20 @@
...
     void SetUp() override {
         tree = gobot::SceneTree::New<gobot::SceneTree>();
+        tree->Initialize();
         node1 = gobot::Node::New<gobot::Node>();
         node2 = gobot::Node::New<gobot::Node>();
         node1_1 = gobot::Node::New<gobot::Node>();
     }
 
     void TearDown() override {
+        // TODO(wqq): Delete node directly is not right when node already on the scene tree
```
Comment (StephenYu-R) [inline tests/scene/test_node_complex.cpp:24]: The editor should finalize the root node (WIndow) and delete it, which automatically reset children. While just for tests, Delete is specifically called to delete a child node.

**Review by StephenYu-R — CHANGES_REQUESTED** [review]

File: include/gobot/core/events/event.hpp [hunk 4]
```diff
@@ -40,7 +37,7 @@ enum class EventType
 enum EventCategory
 {
     None                     = 0,
-    EventCategoryApplication = 1 << 0,
+    EventCategoryWindow = 1 << 0,
```
Comment (StephenYu-R) [inline include/gobot/core/events/event.hpp:40]: Please fix alignment

File: include/gobot/scene/scene_tree.hpp [hunk 5]
```diff
@@ -41,10 +51,17 @@ class GOBOT_EXPORT SceneTree : public Object {
...
+    void OnWindowResize(WindowResizeEvent& e);
+
+private:
+    bool quit_ = false;
+
+    Window *root = nullptr;
     int node_count = 0;
 
-    static SceneTree *singleton;
+    static SceneTree *s_singleton;
```
[the remark this answers is not in the document]
Comment (Wuqiqi123) [reply include/gobot/scene/scene_tree.hpp:64]: > s_singleton_

all static variable has no postfix "_"

**Review by StephenYu-R — CHANGES_REQUESTED** [review]

File: src/gobot/scene/scene_tree.cpp (hunk 1 above)
Comment (Wuqiqi123) [reply src/gobot/scene/scene_tree.cpp:39]: Done

File: tests/scene/test_node_complex.cpp (hunk 3 above)
Comment (Wuqiqi123) [reply tests/scene/test_node_complex.cpp:24]: Delete SceneTree or SceneTree::Fnitialize may automaticlly delete Node you new here, which can cause double free if you delete node afterward.

**Review by StephenYu-R — APPROVED** [review]
LGTM.

**Pull request merged**